docs/proposal/20240308-vpc-infra-creation.md

```diff
@@ -0,0 +1,456 @@
+# Dynamically create infrastructure required for VPC cluster
+
+## Motivation
+Currently, inorder to create a VPC cluster using cluster api we need to create
+resources as prerequisites. This involves work which is limiting the true
+capabilities of cluster api.
+
+
+## Goal
+1. Dynamically create the required cloud resources as a part of cluster
+   creation.
+2. Allow users to access the cluster via loadbalancer, using an IBM Cloud
+   DNS provider.
+
+
+## Proposal
+
+
+### Cluster API VPC cluster components
+
+![vpc-cluster-components.png](../images/vpc-cluster-components.png)
+
+
+### Proposed API changes
+
+
+#### VPCCluster improvements
+```go
+// IBMVPCClusterSpec defines the desired state of IBMVPCCluster.
+type IBMVPCClusterSpec struct {
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// cisInstance defines the IBM Cloud CIS instance to create DNS Records for the cluster.
+	// +optional
+	CISInstance *CISInstance `json:"cisInstance,omitempty"`
+
+	// controlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint capiv1beta1.APIEndpoint `json:"controlPlaneEndpoint"`
+
+	// controlPlaneLoadBalancer is optional configuration for customizing control plane behavior.
+	// +optional
+	ControlPlaneLoadBalancer []*VPCLoadBalancerSpec `json:"controlPlaneLoadBalancer,omitempty"`
+
+	// cosInstance is the IBM COS instance to use for cluster resources.
+	COSInstance *CosInstance `json:"cosInstance,omitempty"`
+
+	// dnsServicesInstance defines the IBM Cloud DNS Services instance to create DNS Records for the cluster.
+	// +optional
+	DNSServicesInstance *DNSServicesInstance `json:"dnsServicesInstance,omitempty"`
```

Comment:
Usually DNS is handled by the openshift itself for other providers like powervs, aws etc.. any more information of the need for adding DNS will be helpful for further discussions

Reply:
Sure, I will remove this then, with that expectation.

Reply:
Removed references.